### ISSUE TITLE
Implied alignment implementation

### ISSUE THREAD

**Issue by recursion-ninja** [amnh/PCG#64]
Implement implied alignment algorithm. It's already implement with the old data types. Some type changes are required and simplifying the logic for a post order & pre-order traversal.

**Comment by recursion-ninja** [comment]
We'll need to encode an `AlignmentContext` value in each of the `DynamicCharacter` streams whch replace the current `DynamicCharacterElement` values.

Something like this:

    data  AlignmentContext
        = Align  x y z
        | Delete x   z
        | Insert   y z

However, this representation will need to be "bit-packed" into the dynamic character stream so that it can be represented by a `BitMatrix`. This can be done by using 2 prefix bits to uniquely identify which context the element is from, and then 3x the number of bits for a single `DynamicCharacterElement` to store `x`, `y`, & `z`. Note that in the `Delete` and `Insert` value representations, the `y` and `z` values will be represented by a "gap" character, respectively.

We should be able coerce the output from the C code's FFI to this format by zipping over the three aligned sequences and interpreting their results accordingly.

**Comment by recursion-ninja** [comment]
**Note:** Turns out that we **cannot** coerce the output from the C code's FFI to this format. There are alignments which return sequences from which it cannot not be infer which output elements correspond to which input elements.

The implied alignment pre-order code has been added. See this commit: fd49df4f20c2a9cfabc0c1f7ec944b4cdbf494e6.

This will be merged into `master` after more work has been done optimizing the dynamic character post-order code.

**Comment by wardwheeler** [comment]
As you feared.
W

On Feb 26, 2020, at 7:50 AM, recursion-ninja <notifications@github.com<mailto:notifications@github.com>> wrote:

EXTERNAL SENDER


Note: Turns out that we cannot coerce the output from the C code's FFI to this format. There are alignments which return sequences from which it cannot not be infer which output elements correspond to which input elements.

The implied alignment pre-order code has been added. See this commit: fd49df4<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Famnh%2FPCG%2Fcommit%2Ffd49df4f20c2a9cfabc0c1f7ec944b4cdbf494e6&data=02%7C01%7Cwheeler%40amnh.org%7Cb462730760784fa3767108d7baba66c9%7Cbe0003e8c6b9496883aeb34586974b76%7C0%7C0%7C637183182075324821&sdata=1%2BDVdmDT0kIfk03BCqE%2Fk5kSn6DKJIFI7IJ7u0Ymz8w%3D&reserved=0>.

This will be merged into master after more work has been done optimizing the dynamic character post-order code.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Famnh%2FPCG%2Fissues%2F64%3Femail_source%3Dnotifications%26email_token%3DABJAZZHNPYXHW4OHURWVZG3REZQPVA5CNFSM4ERCMEV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENADDTQ%23issuecomment-591409614&data=02%7C01%7Cwheeler%40amnh.org%7Cb462730760784fa3767108d7baba66c9%7Cbe0003e8c6b9496883aeb34586974b76%7C0%7C0%7C637183182075324821&sdata=SmLaxFP9kwnD6pfMZggIapPeEQS5vTZcKDsPKnGrSno%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABJAZZDC5GZ5K4TMTH6HWV3REZQPVANCNFSM4ERCMEVQ&data=02%7C01%7Cwheeler%40amnh.org%7Cb462730760784fa3767108d7baba66c9%7Cbe0003e8c6b9496883aeb34586974b76%7C0%7C0%7C637183182075334814&sdata=thaayzMavGpJrU2TaHnMJr%2F5Zr1F%2FqHnxiopeKfB3OA%3D&reserved=0>.

Ward Wheeler
Division of Invertebrate Zoology
American Museum of Natural History
Central Park West at 79th Street
New York, NY 10024-5192
USA
wardwheeler.wordpress.com<http://wardwheeler.wordpress.com>

Google Scholar:
https://scholar.google.com/citations?user=Rjw5orwAAAAJ&hl=en

POY5 Downloads:
https://www.dropbox.com/sh/gb6n6efd4w3hwsj/AACVjPPAT75fBiqmAIrf0qIoa?dl=0

**Comment by recursion-ninja** [comment]
I came up with a solution to converting the C results to our Haskell data type. I had to modify the `backtrace` code and the `get_medians` code on the C side.

In the `backtrace` section, instead of adding a `gap_char` on the traceback when a leftward or upward arrow was encountered, we instead added a `'\0'` value (no bits set). Since `gap_char` has one bit set at the gap character index and `'\0'` has no bits set, we can distinguish between "input gaps" and "output gaps" across the FFi.

The `get_medians` section had to be modified as well, to not look for `gap_char` in the alignments but instead look for `'\0'` and then correctly generate the median for `sigma(x, -)` when `(x, '\0')` was encountered.

The C code is now compatible with the Haskell data-types and the C pairwise alignment returns the same results as the Haskell pairwise alignments.

**Comment by wardwheeler** [comment]
Sweet

Sent from mobile


On Apr 4, 2020, at 2:18 PM, recursion-ninja <notifications@github.com> wrote:

﻿
EXTERNAL SENDER


I came up with a solution to converting the C results to our Haskell data type. I had to modify the backtrace code and the get_medians code on the C side.

In the backtrace section, instead of adding a gap_char on the traceback when a leftward or upward arrow was encountered, we instead added a '\0' value (no bits set). Since gap_char has one bit set at the gap character index and '\0' has no bits set, we can distinguish between "input gaps" and "output gaps" across the FFi.

The get_medians section had to be modified as well, to not look for gap_char in the alignments but instead look for '\0' and then correctly generate the median for sigma(x, -) when (x, '\0') was encountered.

The C code is now compatible with the Haskell data-types and the C pairwise alignment returns the same results as the Haskell pairwise alignments.
